src/system/SystemPacketBuffer.h

```diff
@@ -278,22 +448,96 @@ class DLL_EXPORT PacketBufferHandle
+     */
     PacketBufferHandle Retain() const
     {
         mBuffer->AddRef();
         return PacketBufferHandle(mBuffer);
     }
 
+    /**
+     * Access a PackerBuffer's public methods.
+     */
     PacketBuffer * operator->() const { return mBuffer; }
     PacketBuffer & operator*() const { return *mBuffer; }
 
-    // The caller's ownership is transferred to this.
+    /**
+     * Test whether this PacketBufferHandle is empty, or owns a PacketBuffer.
+     *
+     * @return \c true if this PacketBufferHandle is empty; return \c false if it owns a PacketBuffer.
+     */
+    bool IsNull() const { return mBuffer == nullptr; }
+
+    /**
+     *  Detach and return the head of a buffer chain while updating this handle to point to the remaining buffers.
+     *  The current buffer must be the head of the chain.
+     *
+     *  This PacketBufferHandle now holds the ownership formerly held by the head of the chain.
+     *  The returned PacketBufferHandle holds the ownership formerly held by this.
+     *
+     *  @return the detached buffer formerly at the head of the buffer chain.
+     */
+    CHECK_RETURN_VALUE PacketBufferHandle PopHead();
+
+    /**
+     * Free the first buffer in a chain.
+     *
+     *  @note When the buffer chain is referenced by multiple handles, `FreeHead()` will detach the head, but will not forcibly
+     *  deallocate the head buffer.
+     */
+    void FreeHead()
+    {
+        // `PacketBuffer::FreeHead()` frees the current head; this takes ownership from the `next` link.
+        mBuffer = PacketBuffer::FreeHead(mBuffer);
+    }
+
+    /**
+     * Consume data in a chain of buffers.
+     *
+     *  Consume data in a chain of buffers starting with the current buffer and proceeding through the remaining buffers in the
+     *  chain. Each buffer that is completely consumed is freed and the handle holds the first buffer (if any) containing the
+     *  remaining data. The current buffer must be the head of the buffer chain.
+     *
+     *  @param[in] aConsumeLength - number of bytes to consume from the current chain.
+     */
+    void Consume(uint16_t aConsumeLength) { mBuffer = mBuffer->Consume(aConsumeLength); }
+
+    /**
+     * Copy the given buffer to a right-sized buffer if applicable.
+     * This function is a no-op for sockets.
+     *
+     *  @param[in] aPacket - buffer or buffer chain.
+     *
+     *  @return new packet buffer or the original buffer
+     */
+    void RightSize()
+    {
+#if CHIP_SYSTEM_CONFIG_USE_LWIP && LWIP_PBUF_FROM_CUSTOM_POOLS
+        mBuffer = PacketBuffer::RightSize(mBuffer);
+#endif
+    }
+
+    /**
+     * Take ownership of a raw PacketBuffer pointer.
+     *
+     * @brief The caller's ownership is transferred to this. An existing owned buffer is freed.
+     *
+     * @note This should only be used in low-level code, e.g. to import buffers from LwIP or a similar stack.
+     */
     void Adopt(PacketBuffer * buffer)
```

Comment:
I'm thinking of purging this, even internal to `PacketBufferHandle`, because I've more than once made errors due to the asymmetry: it `Free`s the old `mBuffer` but does not `AddRef` the new one.